src/Utils.vala

```diff
@@ -117,4 +118,140 @@ namespace Utils {
+
+    private static string? get_sys_vendor () {
+        string vendor;
+        try {
+            FileUtils.get_contents ("/sys/devices/virtual/dmi/id/sys_vendor", out vendor);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return vendor.strip ();
+    }
+
+    private static string? get_product_name () {
+        string model;
+        try {
+            FileUtils.get_contents ("/sys/devices/virtual/dmi/id/product_name", out model);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return model.strip ();
+    }
+
+    private static string? get_product_version () {
+        string model;
+        try {
+            FileUtils.get_contents ("/sys/devices/virtual/dmi/id/product_version", out model);
+        } catch (FileError e) {
+            warning ("%s", e.message);
+            return null;
+        }
+
+        return model.strip ();
+    }
+
+    // Based on https://git.launchpad.net/ubiquity/tree/ubiquity/misc.py?id=ae6415d224c2e76afa2274cc9f85997f38870419#n648
+    private static string? get_ubiquity_compatible_hostname () {
+        string model = get_product_name ();
+        string manufacturer = get_sys_vendor ();
+
+        if (manufacturer.length == 0) {
+            return null;
+        }
+        manufacturer = manufacturer.down ();
+
+        if (manufacturer.contains ("to be filled")) {
+            // Don't bother with products in development.
+            return null;
+        }
+
+        if (manufacturer.contains ("bochs") || manufacturer.contains ("vmware")) {
+            model = "virtual machine";
+            // VirtualBox sets an appropriate system-product-name.
+        } else {
+            if (manufacturer.contains ("lenovo") || manufacturer.contains ("ibm")) {
+                model = get_product_version ();
+            }
+        }
+
+        try {
+            if (manufacturer.contains ("apple")) {
+                //  MacBook4,1 - strip the 4,1
+                var re = new Regex ("[^a-zA-Z\\s]");
+                model = re.replace (model, model.length, 0, "");
+            }
+
+            // Replace each gap of non-alphanumeric characters with a dash.
+            // Ensure the resulting string does not begin or end with a dash.
+            var re = new Regex ("[^a-zA-Z0-9]+");
+            model = re.replace (model, model.length, 0, "-");
+            while (model[0] == '-') {
+                model = model.substring (1);
+            }
+            while (model[model.length - 1] == '-') {
+                model = model.substring (0, model.length - 1);
+            }
+
+            if (model.down () == "not-available") {
+                return null;
+            }
+            if (model.down () == "To be filled by O.E.M.".down ()) {
+                return null;
+            }
+        } catch (RegexError e) {
+            print ("Error: %s\n", e.message);
```

Comment:
Would be better to use `warning` here, and remove the `\n`. It would also be preferable to make the error message more clear so it means something in the installer logs (e.g. "Error cleaning up hostname strings: %s"). Probably also want to return null here too.